libs/blog/shared/feature-seo/src/lib/seo/seo.service.ts

```diff
@@ -0,0 +1,219 @@
+import { inject, Injectable } from '@angular/core';
+import { toObservable } from '@angular/core/rxjs-interop';
+import { Meta, MetaDefinition, Title } from '@angular/platform-browser';
+import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
+import { combineLatest, filter, map, switchMap } from 'rxjs';
+
+import { ArticleDetailsStore } from '@angular-love/blog/articles/data-access';
+import { ArticleSeoData } from '@angular-love/contracts/articles';
+
+import { SEO_META_KEYS, SeoMetaKeys } from './seo-meta-keys';
+import { SeoTitleKeys } from './seo-title-keys';
+
+@Injectable()
+export class SeoService {
+  private readonly _articleDetailsStore = inject(ArticleDetailsStore);
+  private readonly _router = inject(Router);
+  private readonly _activatedRoute = inject(ActivatedRoute);
+  private readonly _title = inject(Title);
+  private readonly _meta = inject(Meta);
+
+  constructor() {
+    this.subscribe();
+  }
+
+  private setGlobalSeo(seoData: ArticleSeoData | undefined): void {
+    this.updateTag(seoData?.og_locale || 'en_US', 'ogLocale');
```

Comment:
create an injection token to accept default data, so it won't hold static values here.


```ts
export type SeoConfig = { };
export const SEO_CONFIG = new InjectionToken<SeoConfig>('SEO_CONFIG');
export const provideSeo = (config?: SeoConfig) => {
  return makeEnvironmentProvides([
    {
      provide: SEO_CONFIG,
      useValue: config,
    },
    SeoService
  ])
}
```

Reply:
Done